core/src/credential.rs

```diff
@@ -0,0 +1,82 @@
+
+#![allow(unused_variables)]
+#![allow(dead_code)]
+use std::{time::{SystemTime}};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use crate::HashMap;
+
+// cred_subject is a generic that implements trait X
+// trait X allows us to encode that object into JSON-LD
+// We provide types that implement trait X for the cred types that we support
+// Users can also user their own types that implement trait X if they need a different structure
+// --- 
+// Default context and Cred types are defaulted but can be redefined 
+
+type VerificationContext = [&'static str];
+
+pub const CONTEXT_CREDENTIALS:  VerificationContext = ["https://www.w3.org/2018/credentials/v1", "https://www.w3.org/2018/credentials/examples/v1"];
+
+pub const CRED_TYPE_PERMANENT_RESIDENT_CARD: str = "PermanentResidentCard";
+pub const CRED_TYPE_BANK_CARD: str = "BankCard";
+
+#[derive(Serialize, Deserialize, Clone)]
+struct CredentialSubject {
+    #[serde(flatten)]
+    pub property_set: HashMap<String, Value>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Credential <'a> {
+    #[serde(flatten)]
+    verifiable_credential:VerifiableCredential <'a>,
```

Comment:
I don't think this is a good approach from a strongly typed situation. I think it would be too difficult to utilize. All of the properties should be defined here as normally. I think we should have three types

- Credential - All of the standard properties we defined (non-flattened) and then one property called "custom_properties" or something like that which does get flattened
- IntegrityProof - Same as what we currently have as CredentialProof. Just a more generic name that applies to both VCs and VPs
- Verifiable Credential - The union of the Credential and IntegrityProof - https://doc.rust-lang.org/reference/items/unions.html.

Reply:
Sorry didn't get the idea of union usage. Moreover, to me it looks like we or lib clients have to use unsafe mode since: 

> Consequently, all reads of union fields have to be placed in unsafe blocks: https://doc.rust-lang.org/reference/items/unions.html

Reply:
We definitely want to avoid using unsafe mode if at all possible

Reply:
My bigger issue is with nesting the VerifiableCredential type inside of the Credential type

Reply:
If anything it should be the reverse but I'd rather not do it at all, I think its unnecessarily confusing and seems like a bit of a hack around the limitation of union types